File: storage/result_storage.py
"""
Result Storage - Store and retrieve check results
"""
from typing import Dict, Any, Optional
from datetime import datetime, timedelta
import json


class ResultStorage:
    """In-memory storage for check results (can be replaced with Redis)"""
    
    def __init__(self):
        self.results: Dict[str, Dict[str, Any]] = {}
        self.expiry_time = timedelta(hours=24)  # Results expire after 24 hours
    
    def store_result(self, request_id: str, check_type: str, host: str, 
                    nodes: Dict, results: Dict[str, Any]):
        """Store check result"""
        self.results[request_id] = {
            "check_type": check_type,
            "host": host,
            "nodes": nodes,
            "results": results,
            "created_at": datetime.now().isoformat(),
            "expires_at": (datetime.now() + self.expiry_time).isoformat()
        }
    
    def get_result(self, request_id: str) -> Optional[Dict[str, Any]]:
        """Get check result by request ID"""
        if request_id not in self.results:
            return None
        
        result = self.results[request_id]
        
        # Check if expired
        expires_at = datetime.fromisoformat(result["expires_at"])
        if datetime.now() > expires_at:
            del self.results[request_id]
            return None
        
        return result
    
    def get_extended_result(self, request_id: str) -> Optional[Dict[str, Any]]:
        """Get extended check result"""
        result = self.get_result(request_id)
        if not result:
            return None
        
        return {
            "command": result["check_type"],
            "created": int(datetime.fromisoformat(result["created_at"]).timestamp()),
            "host": result["host"],
            "results": result["results"]
        }
    
    def cleanup_expired(self):
        """Remove expired results"""
        now = datetime.now()
        expired_ids = [
            request_id
            for request_id, result in self.results.items()
            if datetime.fromisoformat(result["expires_at"]) < now
        ]
        
        for request_id in expired_ids:
            del self.results[request_id]

File: src/storage/result-storage.ts
/**
 * Result Storage - Store and retrieve check results using PostgreSQL
 */
import prisma from '../database/prisma';
import { generateRequestId } from '../utils/helpers';

export class ResultStorage {
  async storeResult(
    requestId: string,
    checkType: string,
    host: string,
    nodes: Record<string, string[]>,
    results: Record<string, any>
  ): Promise<void> {
    const expiresAt = new Date();
    expiresAt.setHours(expiresAt.getHours() + 24); // Expire after 24 hours

    try {
      // Create check request
      await prisma.checkRequest.create({
        data: {
          requestId,
          checkType,
          host,
          permanentLink: `${process.env.BASE_URL || 'http://localhost:8000'}/check-report/${requestId}`,
          expiresAt
        }
      });

      // Store results for each node
      for (const [nodeId, resultData] of Object.entries(results)) {
        // Extract agent ID from node ID (e.g., "us1.node.check-host.net" -> "us1")
        const agentId = nodeId.split('.')[0];

        await prisma.checkResult.create({
          data: {
            requestId,
            agentId,
            nodeId,
            resultData: resultData as any,
            status: 'completed'
          }
        });
      }
    } catch (error) {
      console.error('Error storing result:', error);
      throw error;
    }
  }

  async getResult(requestId: string): Promise<any> {
    try {
      const request = await prisma.checkRequest.findUnique({
        where: { requestId },
        include: {
          results: true
        }
      });

      if (!request) {
        return null;
      }

      // Check if expired
      if (new Date() > request.expiresAt) {
        await this.deleteResult(requestId);
        return null;
      }

      // Format results
      const results: Record<string, any> = {};
      for (const result of request.results) {
        results[result.nodeId] = result.resultData;
      }

      return results;
    } catch (error) {
      console.error('Error getting result:', error);
      return null;
    }
  }

  async getExtendedResult(requestId: string): Promise<any> {
    const request = await prisma.checkRequest.findUnique({
      where: { requestId },
      include: {
        results: true
      }
    });

    if (!request) {
      return null;
    }

    // Check if expired
    if (new Date() > request.expiresAt) {
      await this.deleteResult(requestId);
      return null;
    }

    // Format extended results
    const results: Record<string, any> = {};
    for (const result of request.results) {
      results[result.nodeId] = result.resultData;
    }

    return {
      command: request.checkType,
      created: Math.floor(request.createdAt.getTime() / 1000),
      host: request.host,
      results
    };
  }

  async deleteResult(requestId: string): Promise<void> {
    try {
      await prisma.checkRequest.delete({
        where: { requestId }
      });
    } catch (error) {
      console.error('Error deleting result:', error);
    }
  }

  async cleanupExpired(): Promise<void> {
    try {
      const now = new Date();
      await prisma.checkRequest.deleteMany({
        where: {
          expiresAt: {
            lt: now
          }
        }
      });
    } catch (error) {
      console.error('Error cleaning up expired results:', error);
    }
  }
}

